### PR TITLE
Set grep tool

### PR DESCRIPTION
**Pull request by Zarathustra2**
Hey there,

first of all. I really enjoy the plugin, thanks for all the effort you have put in. It is truly beautiful!

This is a follow up of #86 . 

Users are now able to set their grep_tool and set the complete parameters for their grep_tool.

I was thinking if I should always pass the --vimgrep flag to the tool regardless of the tool. If someone has set g:clap_provider_grep_opts like #86 he would have to change it to pass the --vimgrep flag. What are your thoughts?

Best regards

### CODE REVIEW

File: doc/clap.txt [hunk 1]
```diff
@@ -390,6 +390,16 @@ The option naming convention for provider is `g:clap_provider_{provider_id}_{opt
 -------------------------------------------------------------------------------
 6.1. Clap Grep Options                                      *clap-grep-options*
 
+g:clap_provider_grep_cmd                          *g:clap_provider_grep_cmd*
```
Comment (liuchengxu) [inline doc/clap.txt]: How about `g:clap_provider_grep_executable`?

File: autoload/clap/provider/grep.vim [hunk 2]
```diff
@@ -29,11 +30,11 @@ function! s:draw_icon(line) abort
...
 
 function! s:cmd(query) abort
-  if !executable('rg')
-    call g:clap.abort('rg not found')
+  if !executable(s:grep_executable)
+    call g:clap.abort(s:grep_executable . ' not found')
     return
   endif
-  let cmd = 'rg -H --no-heading --vimgrep --smart-case '.s:grep_opts.' "'.a:query.'"'.(has('win32') ? ' .' : '')
+  let cmd = s:grep_executable.' '.s:grep_opts.' "'.a:query.'"'.(has('win32') ? ' .' : '')
```
Comment (liuchengxu) [inline autoload/clap/provider/grep.vim]: I'm not sure if `(has('win32') ? ' .' : '')` is also required for the other tools.

**Comment by liuchengxu** [comment]
I don't think we should always pass the `--vimgrep` flag.

File: autoload/clap/provider/grep.vim (hunk 2 above)
Comment (Zarathustra2) [reply autoload/clap/provider/grep.vim]: Me neither, I dont have any knowledge about windows :(

File: doc/clap.txt [hunk 3]
```diff
@@ -411,7 +421,7 @@ g:clap_provider_grep_blink                         *g:clap_provider_grep_blink*
 g:clap_provider_grep_opts                          *g:clap_provider_grep_opts*
 
   Type: |String|
-  Default: `''`
+  Default: `'-H --no-heading --vimgrep --smart-case'`
 
   Set this to control the extra arguments that get passed to the grep tool.
   Setting `'--hidden -g "!.git/"'` for example would enable grepping hidden
```
Comment (liuchengxu) [inline doc/clap.txt]: Perhaps also change to `'-H --no-heading --vimgrep --smart-case --hidden -g "!.git/"'`?

Comment (Zarathustra2) [reply doc/clap.txt]: All right, I will adjust it! Maybe, I should add another sentence about the --vimgrep flag?

Comment (liuchengxu) [reply doc/clap.txt]: Sure, that would be better.

File: autoload/clap/provider/grep.vim (hunk 2 above)
Comment (liuchengxu) [reply autoload/clap/provider/grep.vim]:
```diff
--- a/autoload/clap/provider/grep.vim
+++ b/autoload/clap/provider/grep.vim
@@ -8,6 +8,7 @@ let s:grep_delay = get(g:, 'clap_provider_grep_delay', 300)
 let s:grep_blink = get(g:, 'clap_provider_grep_blink', [2, 100])
 let s:grep_opts = get(g:, 'clap_provider_grep_opts', '-H --no-heading --vimgrep --smart-case')
 let s:grep_executable = get(g:, 'clap_provider_grep_executable', 'rg')
+let s:grep_cmd_format = get(g:, 'clap_provider_grep_cmd_format', '%s %s "%s"'.(has('win32') ? ' .' : ''))

 let s:old_query = ''
 let s:grep_timer = -1
@@ -34,7 +35,7 @@ function! s:cmd(query) abort
     call g:clap.abort(s:grep_executable . ' not found')
     return
   endif
-  let cmd = s:grep_executable.' '.s:grep_opts.' "'.a:query.'"'.(has('win32') ? ' .' : '')
+  let cmd = printf(s:grep_cmd_format, s:grep_executable, s:grep_opts, a:query)
   let g:clap.provider.cmd = cmd
   return cmd
 endfunction
```

Adding a new option `g:clap_provider_grep_cmd_format` should be a solution.

Do you mind applying this change in your change and add some doc for it?

Comment (Zarathustra2) [reply autoload/clap/provider/grep.vim]: Hey, 

I have added the new option and updated the docs. Are the docs understandable?

File: doc/clap.txt (hunk 3 above)
Comment (Zarathustra2) [reply doc/clap.txt]: Done!

File: doc/clap.txt [hunk 4]
```diff
@@ -411,12 +421,12 @@ g:clap_provider_grep_blink                         *g:clap_provider_grep_blink*
 g:clap_provider_grep_opts                          *g:clap_provider_grep_opts*
 
   Type: |String|
-  Default: `''`
+  Default: `'-H --no-heading --vimgrep --smart-case --hidden -g "!.git/"'`
```
Comment (liuchengxu) [inline doc/clap.txt]:
```suggestion
  Default: `'-H --no-heading --vimgrep --smart-case'`
```

File: doc/clap.txt [hunk 5]
```diff
@@ -411,12 +421,12 @@ g:clap_provider_grep_blink                         *g:clap_provider_grep_blink*
...
 
-  Set this to control the extra arguments that get passed to the grep tool.
-  Setting `'--hidden -g "!.git/"'` for example would enable grepping hidden
```
Comment (liuchengxu) [inline doc/clap.txt]: What I actually mean is to keep this sentence but replacing `'--hidden -g "!.git/"'` with `'-H --no-heading --vimgrep --smart-case --hidden -g "!.git/"'`.

File: autoload/clap/provider/grep.vim [hunk 6]
```diff
@@ -6,7 +6,9 @@ set cpoptions&vim
 
 let s:grep_delay = get(g:, 'clap_provider_grep_delay', 300)
 let s:grep_blink = get(g:, 'clap_provider_grep_blink', [2, 100])
-let s:grep_opts = get(g:, 'clap_provider_grep_opts', '')
+let s:grep_opts = get(g:, 'clap_provider_grep_opts', '-H --no-heading --vimgrep --smart-case --hidden -g "!.git/"')
```
Comment (liuchengxu) [inline autoload/clap/provider/grep.vim]:
```suggestion
let s:grep_opts = get(g:, 'clap_provider_grep_opts', '-H --no-heading --vimgrep --smart-case')
```

File: doc/clap.txt (hunk 5 above)
Comment (Zarathustra2) [reply doc/clap.txt]: Ohhhhh, completely misunderstood you! Yeah that makes totally sense. Fixed!

File: autoload/clap/provider/grep.vim (hunk 6 above)
Comment (Zarathustra2) [reply autoload/clap/provider/grep.vim]: Fixex!

File: doc/clap.txt (hunk 4 above)
Comment (Zarathustra2) [reply doc/clap.txt]: Fixed!

**Comment by liuchengxu** [comment]
Thanks!

**Pull request merged**